components/net/filemanager_thread.rs

```diff
@@ -108,69 +108,161 @@ enum FileImpl {
 
 struct FileManager<UI: 'static + UIProvider> {
     receiver: IpcReceiver<FileManagerThreadMsg>,
-    store: HashMap<Uuid, FileStoreEntry>,
+    store: Arc<RwLock<FileManagerStore<UI>>>,
     classifier: Arc<MimeClassifier>,
-    ui: &'static UI,
 }
 
 impl<UI: 'static + UIProvider> FileManager<UI> {
     fn new(recv: IpcReceiver<FileManagerThreadMsg>, ui: &'static UI) -> FileManager<UI> {
         FileManager {
             receiver: recv,
-            store: HashMap::new(),
+            store: Arc::new(RwLock::new(FileManagerStore::new(ui))),
             classifier: Arc::new(MimeClassifier::new()),
-            ui: ui,
         }
     }
 
     /// Start the file manager event loop
     fn start(&mut self) {
         loop {
+            let store = self.store.clone();
             match self.receiver.recv().unwrap() {
-                FileManagerThreadMsg::SelectFile(filter, sender, origin) => self.select_file(filter, sender, origin),
-                FileManagerThreadMsg::SelectFiles(filter, sender, origin) => self.select_files(filter, sender, origin),
+                FileManagerThreadMsg::SelectFile(filter, sender, origin) => {
+                    spawn_named("select file".to_owned(), move || {
+                        store.write().unwrap().select_file(filter, sender, origin);
```

Comment:
I don't like this, we're locking the store and blocking on the user clicking something. There should never be critical sections that block on user input.

I think we should pass the store in, and within these methods we should lock right where it is necessary, and not more.